### PR TITLE
Structure proposal

### PR DESCRIPTION
**Pull request by x3ro**
Based on our conversation in Discord earlier, here's a proposal for how we could split up the cargo entrypoint and a direct CLI tool, which mirrors how the npm `concurrently` package works. What do you think?

### CODE REVIEW

**Comment by tqwewe** [comment]
Would this structure mean there are two different versions of the cli? One called `$ concurrently ...` and one `$ cargo-concurrently ...`?

I wonder if it would add too much complexity to maintain and deploy two separate clis. I wonder if it can be just one, so the binary would be `src/main.rs`, and lib can be `src/lib.rs`.

I was thinking the following behaviour:

- No arguments, means it will use the `tasks.toml` file in the root of your project to execute commands.
  ```bash
  $ concurrently
  Running tasks from tasks.toml file...
  ```
- Quoted arbitrary arguments, means it will run the provided commands one by one as concurrently on NPM does.
  ```bash
  $ concurrently "command1 arg" "command2 arg"
  [command1] some output
  [command2] some other output
  ```
- Command starting with `"cargo:"` would be treated differently - all cargo commands would be built first, and once all are built then all commands will run. (Kind of like a pre-task, and same behvaiour you can get with tasks.toml file by setting `cargo_workspace_member = true`).
  ```bash
  $ concurrently "cargo:api" "cargo:my-app" "command3"
  Building packags api and my-app...
  [api] Built.
  [my-app] Built.
  Running commands...
  [command3] some output
  ```

What are your thoughts on this approach?

**Comment by x3ro** [comment]
My experience with extending cargo is limited, but as I understand, the idea is similar to git: put `cargo-name-of-your-tool` in the path, which then allows you to execute `cargo name-of-your-tool` as if it was a subcommand ([source](https://doc.rust-lang.org/book/ch14-05-extending-cargo.html#extending-cargo-with-custom-commands)).

I thought this might actually end up being easier than having a single binary, because we wouldn't have to implement two different "modes". Currently, for example, you're taking a lists of `tasks` to be executed, so one could run `cargo concurrently api build`. This would no longer be possible if we also want the same API as the NPM package. You proposed then prefixing, with `cargo:` or the like, but I feel like, as a user of the tool, `cargo concurrently api` would be preferable to `cargo concurrently "cargo:api"` :D

Long story short, my idea is this:

1. You already have a configuration data structure, why we can use as a basis
2. Now we create two CLIs, one for the cargo integration and one for direct usage
    * The cargo subcommand is the one which has the cargo-related features
    * The no-prefix one is kept as close to the `concurrently` NPM package as possible (in terms of CLI API)
3. Both of these CLIs are essentially just configuration loaders.




Untitled-2022-01-20-2201

---

PS: I just realized that you might have never intended it to work as a cargo subcommand. I think I somehow got this from the screenshot in the README, but there you're running `cargo install concurrently`, not, as I somehow initially read it, `cargo concurrently` 😅 So the above might be based on wrong assumptions.

**Comment by x3ro** [comment]
I also just saw that the npm package also seems to have these prefix-like additions (which you probably already know :D). E.g. `concurrently "npm:watch-js"`. I'm honestly not 100% sure if I like that. I guess my hope is that the tool also gains popularity outside of the community that uses cargo, in which cases tight cargo integration may be somewhat confusing 🤔

--- 

I think generally I like the CLI you've proposed above. This proposal was largely based on the assumptions what you'd want to add a cargo subcommand, making it possible to run `cargo concurrently`. If that is not the case, I think we can just start with what you've proposed above and see where it takes us. 

What do you think?

**Comment by tqwewe** [comment]
Ah I see, adding it as a cargo subcommand might be interesting, but I wonder if we can keep it as one binary only. Rather than it having different "modes", it's more like each individual command/task has a different "mode".
- Raw command
- Cargo package
- X programming language package

So like I described before, running concurrently would have a "build" step which all need to finish before running the rest of the commands. So a raw command would have no build commands.. a Cargo package command would have a build ommand of `cargo build`, etc.

Sorry to make things even more confusing, but there would also be the option to run commands after the build step, which would be pre tasks. I've found this to be useful for things like killing processes using a specific port, deleting a directory, etc.

So the lifecycle of running concurrently would be:

- **Build step**: build cargo packages, and in the future x programming language (whatever gets supported).
- **Pre command step**: Allows you to specify raw commands to be run after everything is built, and before tasks begin to run. Useful for killing ports, etc.
- **Run commands step**: Run commands finally, in the case of cargo packages it would just be `./target/build/package` to run it directly.

The benefit to this is that:

- All tasks are built before anything actually starts running.
- No need for separate cli users.. for example what if someone uses cargo, but also wants to run some raw commands as well.. I think it makes sense to just keep it as one.

File: tests/macros.rs [hunk 1]
```diff
@@ -0,0 +1,66 @@
+// Mostly taken from
+// https://github.com/BurntSushi/ripgrep/blob/64ac2ebe0f2fe1c8967e7ec550bc32466bf40a07/tests/macros.rs
+// Licensed under the MIT License
+// Copyright (c) 2015 Andrew Gallant
+
+#[macro_export]
+macro_rules! rgtest {
```
Comment (tqwewe) [inline tests/macros.rs:7]: Could this just be replaced with https://github.com/colin-kiegel/rust-pretty-assertions?

**Comment by tqwewe** [comment]
I also like the idea of having a tasks.toml file. So projects can be cloned, and just run `concurrently` to get everything running. Cargo doesn't have a scripts section as you see in package.json.

**Comment by x3ro** [comment]
I also like the idea of a `tasks.toml` file, though we may want to call it `concurrently.toml` for clearer association? 

---

Being able to influence the running order of commands seems useful indeed. We could implement this on top of a more general system to influence command behaviour through what could be called tags, e.g.

```
$ concurrently "tag1:cat foobar.txt" "tag2,tag3:make all"
# or, more specifically
$ concurrently "pre:rm -rf target/" "pre,cargo:build" "restart:start-server.sh"
```

 Off the top of my head I can think of tags like:

* order
    * `pre`: once before the concurrent tasks run
    * `post`: once after the concurrent tasks run
* behaviour:
    * `restart`: restart the command if it exits
    * `recover`: once, after a command exits

Beyond that, commands would simply be executed in the order in which they are given. I haven't given much thought to how this would be represented in the `.toml` file, but I'm sure there's a decent way.

---

Could you elaborate what the benefit of having an explicit `cargo` prefix would be? If I understand you correctly, you would prefer something like

```
concurrently "cargo:build --all-targets"
```

over something like

```
concurrently "pre:cargo build --all-targets"
```

Do I understand that correctly? Just to be clear, the ordering of the commands would come from the order of specifying it on the CLI / the file.

---

Thinking about this some more, we should also make sure we're not trying to re-implement something like [just](https://github.com/casey/just), i.e. a fully-fledged "dependency-aware" build system. 

As such, I'm wondering if we should leave the file-based specification part, as well as complex ordering guarantees, to something like `just` or `make`, and really focus on the aspect of running commands concurrently.

**Comment by tqwewe** [comment]
> I also like the idea of a tasks.toml file, though we may want to call it concurrently.toml for clearer association?

I agree! This would make more sense.

> Beyond that, commands would simply be executed in the order in which they are given. I haven't given much thought to how this would be represented in the .toml file, but I'm sure there's a decent way.

I wasn't planning for tasks to be run in any order.. I was more so thinking that tasks run "concurrently" at the same time.

So all "pre" tasks are run at the same time... if all succeed then all the tasks are spawned up. And all "post" tasks are run at the same time.

> Could you elaborate what the benefit of having an explicit cargo prefix would be?

Honestly, I was thinking "cargo:api" is just a shortcut for:
```bash
concurrently \
  "pre:cargo build -p api" \
  "./target/build/api"
```

Similar to how npm concurrently has the `npm:api` prefix. It is just an ease of use thing, but I guess it simply exands to the snippet I posted just above.

> Do I understand that correctly? Just to be clear, the ordering of the commands would come from the order of specifying it on the CLI / the file.

As I said, I didnt think concurrently would have any concept of ordering outside of the "pre", normal and "post" tasks.

> Thinking about this some more, we should also make sure we're not trying to re-implement something like just, i.e. a fully-fledged "dependency-aware" build system.
> 
> As such, I'm wondering if we should leave the file-based specification part, as well as complex ordering guarantees, to something like just or make, and really focus on the aspect of running commands concurrently.

Yeah I think keeping it simple is nice. Perhaps just having prefixes of "pre", "post" for lifecycle kind of tasks.. and "cargo", "npm" kind of aliases which expand to the according command. So "npm:api" would just expand to "npm run api"
